File: practice-leetcode/src/main/java/cn/wllsrx/zoe/leetcode/FindErrorNums.java
package cn.wllsrx.zoe.leetcode;

import java.util.Arrays;

/**
 * @author zoe
 * <p>
 * 错误的集合
 * <p>
 * 集合 S 包含从1到n的整数。不幸的是，因为数据错误，导致集合里面某一个元素复制了成了集合里面的另外一个元素的值，导致集合丢失了一个整数并且有一个元素重复。
 * <p>
 * 给定一个数组 nums 代表了集合 S 发生错误后的结果。你的任务是首先寻找到重复出现的整数，再找到丢失的整数，将它们以数组的形式返回。
 * <p>
 * 示例 1:
 * <p>
 * 输入: nums = [1,2,2,4]
 * 输出: [2,3]
 * 注意:
 * <p>
 * 给定数组的长度范围是[2, 10000]。
 * 给定的数组是无序的。
 **/
public class FindErrorNums {

    public static void main(String[] args) {
        System.out.println(Arrays.toString(findErrorNums(new int[]{1, 2, 2, 4})));
    }

    private static int[] findErrorNums(int[] nums) {
        int dup = -1, missing = -1;
        for (int i = 1; i <= nums.length; i++) {
            int count = 0;
            for (int num : nums) {
                if (num == i) {
                    count++;
                }
            }
            if (count == 2) {
                dup = i;
            } else if (count == 0) {
                missing = i;
            }
            if (dup > 0 && missing > 0) {
                break;
            }
        }
        return new int[]{dup, missing};
    }
}

File: practice-scala-base/src/leetcode/FindErrorNums.scala
package leetcode

/**
 * @author zoe
 **/
object FindErrorNums {

  def main(args: Array[String]): Unit = {
    println(findErrorNums(Array(1, 2, 2, 4)).mkString(","))
  }

  def findErrorNums(nums: Array[Int]): Array[Int] = {
    var dup = -1
    var missing = -1
    val length = nums.length
    for (i <- 1 to length) {
      var count = 0
      for (num <- 0 until length) {
        if (nums(num) == i) {
          count += 1
        }
      }
      if (count == 2) {
        dup = i
      } else if (count == 0) {
        missing = i
      }
    }
    Array(dup, missing)
  }

}
